Take advantage of C++17 in scalar_type_util.h

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #6968

I generated a big ugly table because we couldn't make promoteTypes constexpr before we had C++17. Now we have C++17.

Differential Revision: [D66181946](https://our.internmc.facebook.com/intern/diff/D66181946/)

swolchok committed:
Take advantage of C++17 in scalar_type_util.h

I generated a big ugly table because we couldn't make promoteTypes constexpr before we had C++17. Now we have C++17.

Differential Revision: [D66181946](https://our.internmc.facebook.com/intern/diff/D66181946/)

[ghstack-poisoned]

swolchok committed:
reinstate typedefs on "Take advantage of C++17 in scalar_type_util.h"


I generated a big ugly table because we couldn't make promoteTypes constexpr before we had C++17. Now we have C++17.

Differential Revision: [D66181946](https://our.internmc.facebook.com/intern/diff/D66181946/)

[ghstack-poisoned]

dbort reviewed (approved):
Approving optimistically, but please make sure CI jobs run properly after the SEV is mitigated

swolchok committed:
Update on "Take advantage of C++17 in scalar_type_util.h"


I generated a big ugly table because we couldn't make promoteTypes constexpr before we had C++17. Now we have C++17.

Differential Revision: [D66181946](https://our.internmc.facebook.com/intern/diff/D66181946/)

[ghstack-poisoned]

swolchok committed:
Update on "Take advantage of C++17 in scalar_type_util.h"


I generated a big ugly table because we couldn't make promoteTypes constexpr before we had C++17. Now we have C++17.

Differential Revision: [D66181946](https://our.internmc.facebook.com/intern/diff/D66181946/)

[ghstack-poisoned]

swolchok committed:
unstack on "Take advantage of C++17 in scalar_type_util.h"


I generated a big ugly table because we couldn't make promoteTypes constexpr before we had C++17. Now we have C++17.

Differential Revision: [D66181946](https://our.internmc.facebook.com/intern/diff/D66181946/)

[ghstack-poisoned]

swolchok committed:
Update on "Take advantage of C++17 in scalar_type_util.h"


I generated a big ugly table because we couldn't make promoteTypes constexpr before we had C++17. Now we have C++17.

Differential Revision: [D66181946](https://our.internmc.facebook.com/intern/diff/D66181946/)

[ghstack-poisoned]
